File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        nodes = []
        
        ptr = head
        while ptr:
            nodes.append(ptr)
            ptr = ptr.next

        left = 0
        right = len(nodes) - 1

        updated_list = ListNode()
        pointer_to_updated_list = updated_list
        
        while left <= right:
            node_left = nodes[left]
            node_right = nodes[right]
            node_left.next = None
            node_right.next = None
            if left == right:
                updated_list.next = node_left
                break
            else:
                updated_list.next = node_left
                updated_list.next.next = node_right
                updated_list = node_right
            left += 1
            right -= 1
        
        head = pointer_to_updated_list.next


        
        

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) {
            return null;
        }

        PriorityQueue<ListNode> minHeap = new PriorityQueue<>((a, b) -> Integer.compare(a.val, b.val));
        ListNode resultNode = new ListNode();
        ListNode ptr = resultNode;
        
        // Insert all the first nodes to the pq.
        for (ListNode linkedList : lists) {
            if (linkedList != null) {
                minHeap.add(linkedList);
            }
        }

        // Now we keep popping from the linkedlist, and the list which was read from now.
        // we move the pointer to the next of it, and put that in the pq.
        while (!minHeap.isEmpty()) {
            ListNode topElement = minHeap.poll();
            ptr.next = topElement;
            ptr = ptr.next;
            if (topElement.next != null) {
                minHeap.add(topElement.next);
            }
        }

        return resultNode.next;
    }
}
